zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1566,7 +1569,7 @@ private void reconnect() {
     getEventLock().lock();
     try {
       ZkConnection connection = ((ZkConnection) getConnection());
-      connection.reconnect(this);
+      connection.reconnect(_watcher);
```

Comment:
What if the passed in watcher is not null? Then it used the watcher? I am not sure whether this is correct interpretation of watcher.

Reply:
If a non-null watcher is used for ZkClient construction, that external watcher will be used for ZK notification. This is my current idea for using different watcher and event thread.